lib/auth/machineid/machineidv1/bot_service.go

```diff
@@ -440,6 +441,11 @@ func (bs *BotService) UpdateBot(
 	for _, path := range req.UpdateMask.Paths {
 		switch {
 		case path == "spec.roles":
+			if slices.Contains(req.Bot.Spec.Roles, "") {
+				return nil, trace.BadParameter(
+					"spec.roles: must not contain empty strings",
+				)
+			}
```

Comment:
> Is it possible to submit an update request with "spec.roles" == []? If so is it valid to update a bot to have 0 roles? If not should we reject in that scenario?

It's valid for a Bot to have no roles - it's not a super useful Bot at the moment but it may become more useful in future and it gives us a little more flexibility in how the guides are written (e.g in the first half (platform guides) the Bot is created without roles, and the second half (access guides )describe creating a role and assigning it to the Bot).

> Should we filter out empty roles and allow the request to proceed if at least one non-empty role was provided instead of rejecting?

I have a bit of an aversion to these kinds of "graceful" handling of invalid parameters. I find it makes the behaviour much less predictable and can make things much harder to diagnose.